pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -396,9 +411,13 @@ pub(super) async fn prepare(
                 new_layers.push(owned);
             }
             Ok(Err(failed)) => {
+                delete_layers(detached, new_layers)?;
                 return Err(failed);
             }
-            Err(je) => return Err(Error::Prepare(je.into())),
+            Err(je) => {
+                delete_layers(detached, new_layers)?;
+                return Err(Error::Prepare(je.into()));
+            }
```

Comment:
Hm, isn't this thing going to be racy? There may still be in-flight tasks that we haven't gotten a result from, and they may still be writing hardlinks after we return here. Should we cancel all the tasks when we get an error, wait for them all to terminate, and then clean up?

Reply:
cancellation is inherently racy, so we'll probably have to run all tasks to their end.

Reply:
I don't think it is here? It seems like all of the IO operations we're doing are sync, via `std::fs`, and we don't care about the network IO. We should be able to just cancel the tasks and wait for the join handles.

Reply:
I've added a cancellation token instead. tbh now I'm less sure if it's worth to clean up after ourselves or to just rely on the code that makes the hard linking code delete already existing files and retry (also in this PR).

Reply:
Seems polite to clean up after us.